### PR TITLE
PARQUET-528: Fix flush() for RecordConsumer and implementations

### PR DESCRIPTION
**Pull request by proflin**
`flush()` was added in `RecordConsumer` and `MessageColumnIO` to help implementing nulls caching.

However, other `RecordConsumer` implementations should also implements `flush()` properly. For instance, `RecordConsumerLoggingWrapper` and `ValidatingRecordConsumer` should call `delegate.flush()` in their `flush()` methods, otherwise data might be mistakenly truncated.

This PR:
- makes `flush()` abstract in `RecordConsumer`
- implements `flush()` properly for all `RecordConsumer` subclasses, specifically:
 - `RecordConsumerLoggingWrapper`
 - `ValidatingRecordConsumer`
 - `ConverterConsumer `
 - `ExpectationValidatingRecordConsumer `

### CODE REVIEW

**Comment by proflin** [comment]
@julienledem @rdblue @liancheng would you mind taking a look at this when you have time? This has been blocking [[Parquet-401: Deprecate Log and move to SLF4J Logger][PR#319]](https://github.com/apache/parquet-mr/pull/319). Thanks!

File: parquet-column/src/main/java/org/apache/parquet/io/api/RecordConsumer.java [hunk 1]
```diff
@@ -126,10 +126,8 @@
   abstract public void addDouble(double value);
 
   /**
-   * NoOps by default
-   * Subclass class can implement its own flushing logic
+   * subclass class should implement its own flushing logic
    */
-  public void flush() {
-  }
+  abstract public void flush();
```
Comment (julienledem) [inline parquet-column/src/main/java/org/apache/parquet/io/api/RecordConsumer.java]: I think for backward compatibility this needs to not be abstract.

Comment (proflin) [inline parquet-column/src/main/java/org/apache/parquet/io/api/RecordConsumer.java]: @julienledem 
Sure. I've issued another commit reverting it back to be non-_abstract_. Is there any need we make it _abstract_ in a major release, like 2.0?

**Comment by proflin** [comment]
hi @julienledem, sorry for keeping pinging on this, but would you take another look when you have time?
Thanks!

**Comment by rdblue** [comment]
+1, looks fine to me.

@julienledem may want to comment, but otherwise I'll merge this in a day or so.

**Comment by julienledem** [comment]
+1

**Comment by liancheng** [comment]
Thanks, merged to master.

**Pull request closed without merging**